include/Common/SharedQueue.h

```diff
@@ -0,0 +1,90 @@
+#pragma once
+
+#include <queue>
+#include <mutex>
+#include <condition_variable>
+
+template <typename T>
+class SharedQueue
+{
+public:
+    SharedQueue();
+    ~SharedQueue();
+
+    T& front();
+    void pop_front();
+
+    void push_back(const T& item);
+    void push_back(T&& item);
+
+    int size();
+    bool empty();
+
+private:
+    std::deque<T> queue_;
+    std::mutex mutex_;
+    std::condition_variable cond_;
+};
+
+template <typename T>
+SharedQueue<T>::SharedQueue() {}
+
+template <typename T>
+SharedQueue<T>::~SharedQueue() {}
+
+template <typename T>
+T& SharedQueue<T>::front()
```

Comment:
Move the function implementations into the actual class definition. Otherwise, I think you'll have issues if trying to include this header in multiple places.

```
template <typename T>
class SharedQueue
{
public:
    SharedQueue() = default;
    ~SharedQueue() = default;

    T& front()
    {
        // implement here
    }

    void pop_front()
    {
        // implement here
    }

    void push_back(const T& item)
    {
        // implement here
    }

    void push_back(T&& item)
    {
        // implement here
    }

    int size()
    {
        // implement here
    }

    bool empty()
    {
        // implement here
    }

private:
    std::deque<T> queue_;
    std::mutex mutex_;
    std::condition_variable cond_;
};
```